File: problems-adapt/0003_longest-duplicate-free-substring/solution_last_index_jump.py
from typing import List, Optional


class Solution:
    def longestDuplicateFreeLength(self, s: str) -> int:
        # last_seen maps each character to its most recent index; the window
        # s[start..i] is kept duplicate-free throughout the sweep.
        last_seen = {}
        start = 0
        best = 0
        for i, c in enumerate(s):
            # The >= start guard ignores occurrences that lie to the left of
            # the window; without it start could be dragged backwards.
            if c in last_seen and last_seen[c] >= start:
                # The window can no longer include that older occurrence, so
                # start leaps over the conflict instead of shrinking by one.
                start = last_seen[c] + 1
            last_seen[c] = i
            # Invariant restored: the window is duplicate-free again.
            best = max(best, i - start + 1)
        return best

File: problems-originals/0001-0100/0003_longest-substring-without-repeating-characters/solution_last_index_jump.ts
function lengthOfLongestSubstring(s: string): number {
    // last maps each character to its most recent index; the window
    // s[start..i] is kept duplicate-free throughout the sweep.
    const last = new Map<string, number>();
    let start = 0,
        best = 0;
    for (let i = 0; i < s.length; i++) {
        const c = s[i];
        // The >= start guard ignores occurrences left of the window;
        // without it start could be dragged backwards.
        if (last.has(c) && last.get(c)! >= start) {
            // The window can no longer include that older occurrence, so
            // start leaps over the conflict instead of shrinking by one.
            start = last.get(c)! + 1;
        }
        last.set(c, i);
        // Window is duplicate-free again: record its length.
        if (i - start + 1 > best) best = i - start + 1;
    }
    return best;
}
